Use Py_SET_SIZE() on Python 3.9.0a4 and newer

vstinner opened the pull request:
Py_SIZE() must not be used as an l-value anymore in Python 3.9:
Py_SET_SIZE() must be used instead:

* https://bugs.python.org/issue39573
* https://docs.python.org/dev/c-api/structures.html#c.Py_SET_SIZE

vstinner commented:
Maybe a __Pyx_Py_SET_SIZE() wrapper should be added which use use Py_SET_SIZE() if available, or use Py_SIZE() otherwise. But I don't know how to do that.

I don't know if tests/run/verbatiminclude.pyx test was valid or not. In case of doubt, I rename the test function. Does the test require that Py_SET_SIZE() already exists in Python.h? Or does it define a new function?

scoder commented:
Thanks Victor!

> Or does it define a new function?

This. And that's intentional in the test, it aims to define something in C. I think renaming the function there is the best way forward, as you did.

> __Pyx_Py_SET_SIZE()

Yes, I think that would be the safer (and less intrusive) way to do this. There is a `PythonCompatibility` section in `ModuleSetupCode.c`. Just add it there, maybe below the `__Pyx_PySequence_SIZE` define.

vstinner commented:
Ok, I updated my PR to add a new _Pyx_SET_SIZE() function.

I also fixed Cython/Compiler/ModuleNode.py to avoid using Py_REFCNT() as an l-value.

I created this PR to make Cython compatible with my CPython change: https://github.com/python/cpython/pull/20429

scoder commented on Cython/Compiler/ModuleNode.py:
```diff
@@ -1692,9 +1692,9 @@ def generate_usr_dealloc_call(self, scope, code):
+        code.putln("Py_INCREF(o);")
         code.putln("%s(o);" % entry.func_cname)
-        code.putln("--Py_REFCNT(o);")
+        code.putln("Py_DECREF(o);")
```
The thing is, this touches the deallocation machinery, and we're in deallocation code already here. The code pretty intentionally only changes the refcount (to temporarily keep the object alive / resurrect it) for calling into (arbitrary) user defined code, before continuing to clean it up.

vstinner replied:
Oh I see, I modified my PR to add __Pyx_SET_REFCNT() function instead.

scoder commented on Cython/Utility/ModuleSetupCode.c:
```diff
@@ -829,6 +829,13 @@ static CYTHON_INLINE PyObject * __Pyx_PyDict_GetItemStrWithError(PyObject *dict,
 
+
+#if PY_VERSION_HEX >= 0x030900A4
+  #define _Pyx_SET_SIZE(obj, size) Py_SET_SIZE(obj, size)
+#else
+  #define _Pyx_SET_SIZE(obj, size) do { Py_SIZE(obj) = (size); } while (0)
```
The prefix is `__Pyx_`, not `_Pyx_`.
Also, the `do…while` isn't needed since we control the usages on our side.

vstinner replied:
Oops, I fixed both issues.

vstinner commented:
I squashed my commits to rewrite commit messages.

scoder commented on tests/run/verbatiminclude.pyx:
```diff
@@ -25,9 +25,9 @@ cdef class C:
 cdef extern from "Python.h":
     """
-    #define Py_SET_SIZE(obj, size)  Py_SIZE((obj)) = (size)
+    #define my_SET_SIZE(obj, size)  Py_SIZE((obj)) = (size)
```
This test would probably fail after your CPython change, though.
```suggestion
    #define my_SET_SIZE(obj, size)  __Pyx_SET_SIZE((obj), size)
```

scoder commented on Cython/Utility/ModuleSetupCode.c:
```diff
@@ -829,6 +829,20 @@ static CYTHON_INLINE PyObject * __Pyx_PyDict_GetItemStrWithError(PyObject *dict,
 #endif
 
+
+#if PY_VERSION_HEX >= 0x030900A4
+  #define __Pyx_SET_REFCNT(obj, refcnt) Py_SET_REFCNT(obj, refcnt)
+#else
+  #define __Pyx_SET_REFCNT(obj, refcnt) Py_REFCNT(obj) = (refcnt)
+#endif
+
+
+#if PY_VERSION_HEX >= 0x030900A4
+  #define __Pyx_SET_SIZE(obj, size) Py_SET_SIZE(obj, size)
+#else
+  #define __Pyx_SET_SIZE(obj, size) Py_SIZE(obj) = (size)
+#endif
```
Just drop this in one. They are related enough not to bloat C files on user side for them.
```suggestion
#if PY_VERSION_HEX >= 0x030900A4
  #define __Pyx_SET_REFCNT(obj, refcnt)  Py_SET_REFCNT(obj, refcnt)
  #define __Pyx_SET_SIZE(obj, size)      Py_SET_SIZE(obj, size)
#else
  #define __Pyx_SET_REFCNT(obj, refcnt)  Py_REFCNT(obj) = (refcnt)
  #define __Pyx_SET_SIZE(obj, size)      Py_SIZE(obj) = (size)
#endif
```

scoder commented on Cython/Utility/Builtins.c:
```diff
@@ -257,7 +257,7 @@ static PyObject *__Pyx_PyLong_AbsNeg(PyObject *n) {
         PyObject *copy = _PyLong_Copy((PyLongObject*)n);
         if (likely(copy)) {
-            Py_SIZE(copy) = -(Py_SIZE(copy));
+            __Pyx_SET_SIZE(copy, -Py_SIZE(copy));
```
Since this seems a lot less clear now, I think it's worth a comment.
```suggestion
            // negate the size to swap the sign
            __Pyx_SET_SIZE(copy, -Py_SIZE(copy));
```

scoder commented on Cython/Compiler/ModuleNode.py:
```diff
@@ -1692,9 +1692,9 @@ def generate_usr_dealloc_call(self, scope, code):
         code.putln("PyObject *etype, *eval, *etb;")
         code.putln("PyErr_Fetch(&etype, &eval, &etb);")
-        code.putln("++Py_REFCNT(o);")
+        code.putln("__Pyx_SET_REFCNT(o, Py_REFCNT(o) + 1);")
```
```suggestion
        # increase the refcount while we are calling into user code to prevent recursive deallocation
        code.putln("__Pyx_SET_REFCNT(o, Py_REFCNT(o) + 1);")
```

vstinner commented:
@scoder: I updated my PR to address your latest review.

scoder commented:
Ok, let's see if travis gives her placed, then I'll merge this. Thanks Victor!

vstinner commented:
@scoder: Travis CI is happy. The two PyPy jobs failed, but I don't think that it's related to this PR.

vstinner commented:
Thanks for your great advices and also for merging quickly my change! Long life to the new opaque C API :-D

Oh, I tried to backport my fix 0.29.x branch, but you already did, thanks!

vstinner commented:
@scoder: I see that the fix is part of the Cython 0.29.19 milestone https://github.com/cython/cython/milestone/79 but Cython 0.29.19 was released 7 days ago?

Do you plan to release Cython 0.29.20 which will include this fix to make it compatible with future Python 3.10?

scoder commented:
Yes, that was the idea. Looks like I hadn't closed the old milestone yet. Will fix it when I get to it.

vstinner commented:
Hum, https://github.com/cython/cython/commit/54bad41c045c3cc6f0e632ca356283dc147380db#diff-32655df5c2097cca93b52a2ff7583bef merge duplicated the macros. They are now defined twice:
https://github.com/cython/cython/blob/57b4bdeb88cc18f8e2f89fe9a7cf506d4ee0f8db/Cython/Utility/ModuleSetupCode.c#L833

Extract of the code:

```
#if PY_VERSION_HEX >= 0x030900A4
  #define __Pyx_SET_REFCNT(obj, refcnt) Py_SET_REFCNT(obj, refcnt)
  #define __Pyx_SET_SIZE(obj, size) Py_SET_SIZE(obj, size)
#else
  #define __Pyx_SET_REFCNT(obj, refcnt) Py_REFCNT(obj) = (refcnt)
  #define __Pyx_SET_SIZE(obj, size) Py_SIZE(obj) = (size)
#endif


#if PY_VERSION_HEX >= 0x030900A4
  #define __Pyx_SET_REFCNT(obj, refcnt) Py_SET_REFCNT(obj, refcnt)
  #define __Pyx_SET_SIZE(obj, size) Py_SET_SIZE(obj, size)
#else
  #define __Pyx_SET_REFCNT(obj, refcnt) Py_REFCNT(obj) = (refcnt)
  #define __Pyx_SET_SIZE(obj, size) Py_SIZE(obj) = (size)
#endif
```

@scoder: Is it a mistake in a merge?

scoder commented:
Interesting. Yeah, thanks for noticing @vstinner. I removed the duplication in 359f89a8c5250d6dea8da025edbc6e39fbd11885.
